Merge ghc-imported-from into ghc-mod

carlohamalainen opened the pull request:
Merging ghc-imported-from into ghc-mod: https://github.com/carlohamalainen/ghc-imported-from

carlohamalainen committed:
Merging the ghc-imported-from utility.

carlohamalainen committed:
Dependencies for imported-from command.

Also removed the bounds on cabal-helper to build with latest Stack.

carlohamalainen committed:
Add imported-from command.

carlohamalainen committed:
Fix a type error.

carlohamalainen committed:
Remove bound on process-streaming.

carlohamalainen committed:
Fix bounds for haddock-api with ghc 7.x and 8.x

carlohamalainen commented:
Struggling to get this to build on ghc-7.6.3 due to dependencies on process-streaming. Will sort it out.

DanielG commented on ghc-mod.cabal:
```diff
@@ -174,9 +175,10 @@ Library
                       , bytestring        < 0.11
                       , binary            < 0.9 && >= 0.5.1.0
                       , containers        < 0.6
-                      , cabal-helper      < 0.8 && >= 0.7.1.0
```
Don't remove the bounds please :)

carlohamalainen replied:
Reverted ghc-mod.cabal:

```
commit 875b5137a867416bfa7416493a0aac0165c91222
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Fri Jul 15 21:09:59 2016 +1000

    Revert ghc-mod.cabal to current master.
```

DanielG commented on ghc-mod.cabal:
```diff
@@ -193,6 +195,9 @@ Library
                       , mtl               < 2.3 && >= 2.0
                       , monad-control     < 1.1 && >= 1
+                      , parsec
+                      , process
```
Duplicate depdency, thats already declared above.

carlohamalainen replied:
Removed.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+import System.Process.Streaming
+import TcRnTypes()
+
+import qualified Data.ByteString as B.X
```
What's up with this, doesn't seem to be used..?

carlohamalainen replied:
```
commit ffc755b11edf0693856696e91b05962eeb957f1e
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Fri Jul 15 21:18:08 2016 +1000

    Remove unused import.
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+            Left e       -> Nothing
+            Right (a, b) -> Just (b2s a, b2s b)
+      where
+        b2s = map w2c . B.unpack . BL.toStrict
```
Sounds dangerous, I'd rather not use BS.Internal

carlohamalainen replied:
Removed use of `BS.internal` as it was only used along side `System.Process.Streaming`.

```
commit 07e4be3ab967957b24800ba0497ce924e5591c75
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Fri Jul 15 22:43:24 2016 +1000

    Remove dependency on System.Process.Streaming
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+
+    executeFallibly' :: String -> [String] -> IO (Maybe (String, String))
+    executeFallibly' cmd args = do
+        x <- executeFallibly (piped (proc cmd args)) ((,) <$> foldOut intoLazyBytes <*> foldErr intoLazyBytes)
```
S.P.Streaming seems quite pointless for, you could just use `readProcessWithExitCode`. Which makes the whole roundtrip through ByteString pointless too. Given you later just print stderr please just use `gmReadProcess` which returns a function that handles translating stderr for `--line-prefix` stuff.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+        iface <- Haddock.readInterfaceFile nameCacheFromGhc ifile
+
+        case iface of
+            Left _          -> do gmErrStrLn $ "Failed to read the Haddock interface file: " ++ ifile
```
You should use the `MonadError GhcModError` for errors, just add a new constructor to `GhcModError` and add a pretty printer to `gmeDoc`.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+    -- Copied from http://hackage.haskell.org/package/haddock-api-2.16.1/docs/src/Haddock-InterfaceFile.html#nameCacheFromGhc
+    -- but for a general monad m instead of the specific monad Ghc.
+    ------------------------------------------------------------------------------------------------------------------------
+    nameCacheFromGhc :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m) => Haddock.NameCacheAccessor m
```
`GmOut` and `GmLog` seem to be redundant constraints here, you should compile with GHC 8 that warns about those now. Also please don't use `forall` notation if you don't have to.

carlohamalainen replied:
```
commit 7106525731cef1d3082417c00b6afe6dc9aba7bf
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Sun Jul 17 22:03:47 2016 +1000

    Remove constraints on GmOut and GmLog.
```

If I remove the `forall` here it doesn't compile. What's wrong with `forall` notation?

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+           liftIO $ writeIORef ref nc'
+
+-- | Convert a module name string, e.g. @Data.List@ to @Data-List.html@.
+moduleNameToHtmlFile :: String -> String
```
This just has one call site and the type isn't awfully insightful, should probably be in a `where`.

carlohamalainen replied:
```
commit 048eaac9b3c2c431922f8b7a6f6d96b38d7ea53d
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Sun Jul 17 22:12:12 2016 +1000

    Fewer imports; move moduleNameToHtmlFile into a where clause.
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+          f c   = c
+
+filterMatchingQualifiedImport :: String -> [NiceImportDecl] -> [NiceImportDecl]
+filterMatchingQualifiedImport symbol hmodules =
```
Also should be in a `where`, the indenting looks strange also. Newline after `of` please.

carlohamalainen replied:
```
commit 13e5cdabb9bb484c0c843c5752ab07ff559f2366
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Sun Jul 17 22:19:48 2016 +1000

    Move filterMatchingQualifiedImport to a where clause; tidy up
    indenting.
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+    -> m (Maybe ([String], String))
+getModuleExports ghcPkg readProc pkgDbStack m = do
+    minfo     <- (findModule (mkModuleName $ modName m) Nothing >>= getModuleInfo)
+                   `gcatch` (\(_  :: SourceError)   -> return Nothing)
```
There should definitely be some logging happening here.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+        -- both lines and unlines matched. Prepending a dot doesn't work due to things like ".=" from
+        -- Control.Lens. So we manually check that the suffix matches, that the next symbol is a dot,
+        -- and then an alpha character, which hopefully is the end of a module name. Such a mess.
+        where f n qn = if length qn - length n - 2 >= 0
```
this looks very suspicious ;)

DanielG replied:
If you're trying to parse a `hiding` list here maybe you'd be better served by using ReadP rather than this abomination? If you give me some examples what the input looks like I could do it for you.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+    -> [GhcPkgDb]
+    -> m (Either String String)
+guessHaddockUrl modSum targetFile targetModule symbol lineNr colNr ghcPkg readProc pkgDbStack = do
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: targetFile: "   ++ targetFile
```
The argument you're passing as `""` here is what a location qualifier like `"guessHaddockUrl"` should go into.

carlohamalainen replied:
Changed these everywhere.

``````
commit f52077c9afa32164b253a46df236f0fa4a7c4d0f
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Sun Jul 17 22:32:20 2016 +1000

    Use parameter of ```gmLog GmDebug``` instead of passing ```""```.
``````

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+
+        extraImportDecls :: [NiceImportDecl]
+        extraImportDecls = case Safe.headMay parsedPackagesAndQualNames of
+                        Just (Right (_, x)) -> case moduleOfQualifiedName x of Just x' -> [ NiceImportDecl
```
This indent is wayyyy to deep. Keep it below 80 chars where possible please.

carlohamalainen replied:
Fixed.

```
commit 50a52d21e7236815e33f986bb55e851501f1ea6c
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Sun Jul 17 22:38:03 2016 +1000

    Reformat some very long lines.
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+    e <- liftIO $ doesFileExist f
+
+    if e then return $ Right $ "file://" ++ f
+         else do gmErrStrLn "Please reinstall packages using the flag '--enable-documentation' for 'cabal install.\n"
```
This should be a proper top level error in GhcModError too.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+
+          res <- guessHaddockUrl modSum file modstr symbol lineNr colNr ghcPkg readProc pkgDbStack
+
+          case res of Right x  -> return $ "SUCCESS: " ++ x ++ "\n"
```
Success/failure should be returned via the exit code, errors should be printed to stderr.

carlohamalainen replied:
```
commit ae4badbb3984b5f718ff572f46d6b59c18d3a339
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Tue Jul 19 20:06:19 2016 +0800

    Use exit code instead of printing SUCCESS / FAIL.
```

DanielG commented on Language/Haskell/GhcMod/Pretty.hs:
```diff
@@ -32,8 +32,7 @@ gmRenderDoc = renderStyle docStyle
 gmComponentNameDoc ChSetupHsName   = text $ "Setup.hs"
-gmComponentNameDoc (ChLibName "")  = text $ "library"
-gmComponentNameDoc (ChLibName n)   = text $ "library:" ++ n
+gmComponentNameDoc ChLibName       = text $ "library"
```
You're backpatching cabal-helper to <0.7, please don't :)

carlohamalainen replied:
Reverted this one, sorry for the weirdness.

```
commit cfb96931b2317743feab767fe565f5512ba286f5
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Fri Jul 15 22:42:57 2016 +1000

    Revert to current master.

Language/Haskell/GhcMod/Pretty.hs
```

DanielG commented on ghc-mod.cabal:
```diff
@@ -134,6 +134,7 @@ Library
                         Language.Haskell.GhcMod.Flag
                         Language.Haskell.GhcMod.Gap
                         Language.Haskell.GhcMod.GhcPkg
+                        Language.Haskell.GhcMod.ImportedFrom
```
"H" comes before "I", I like to keep this list (semi) sorted.

carlohamalainen replied:
```
commit ae395943d94c340821c277181abdd90b1f5a0d57
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Fri Jul 15 21:13:05 2016 +1000

    Import in alphabetical position.
```

DanielG commented on ghc-mod.cabal:
```diff
@@ -174,9 +175,10 @@ Library
+                      , cabal-helper
                       , deepseq           < 1.5
                       , directory         < 1.3
+                      , exceptions
```
Upper bounds please.

carlohamalainen replied:
This is now

```
exceptions        < 0.9
```

DanielG commented on ghc-mod.cabal:
```diff
@@ -193,6 +195,9 @@ Library
                       , transformers-base < 0.5
                       , mtl               < 2.3 && >= 2.0
                       , monad-control     < 1.1 && >= 1
+                      , parsec
```
Upper bounds too.

carlohamalainen replied:
```
parsec            < 3.2
```

DanielG commented on ghc-mod.cabal:
```diff
@@ -193,6 +195,9 @@ Library
                       , monad-control     < 1.1 && >= 1
+                      , parsec
+                      , process
+                      , process-streaming
```
This doesn't seem to be used for much anything important. I'd rather not depend on it.

carlohamalainen replied:
Removed; not used anymore.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+-- You should have received a copy of the GNU Affero General Public License
+-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+{-# LANGUAGE CPP                    #-}
```
Just to keep it consistent with the rest of the codebase, keep the language extensions in a single comma seperated LANGUAGE declaration. You can still break lines when they become longer than 80 chars.

carlohamalainen replied:
```
commit 47becac075cbce805e3b310fad8a238d3fa51475
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Fri Jul 15 21:16:06 2016 +1000

    LANGUAGE extensions on one line.
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+
+module Language.Haskell.GhcMod.ImportedFrom (importedFrom) where
+
+import Control.Applicative()
```
Are you actually importing instances here? If so mention them in a comment please.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+import qualified SrcLoc
+import qualified Text.Parsec as TP
+
+#if __GLASGOW_HASKELL__ >= 708
```
CPP on the GHC version goes in Gap.hs where possible but you really shouldn't use unsafeGlobalDynFlags anyways.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,785 @@
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: col nr: "       ++ show colNr
+
+    let textualImports  = ms_textual_imps modSum
+        importDecls0 = map toImportDecl textualImports
```
You have a `GhcMonad m` here. Use it for proper pretty printing in toImportDecl instead of using unsafeGlobalDynFlags please. Unless I'm missing a reason everything should always be qualified or something?

carlohamalainen replied:
No reason for it, have removed it.

```
commit d2762605d6ce66523e1018d313a10d7d5a05897d
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Sun Jul 17 21:59:15 2016 +1000

    Remove definition of tdflags (also gets rid of unsafeGlobalDynFlags).
    Just use the DynFlags from the current session.
```

carlohamalainen commented:
Thanks for all the comments; will address all ASAP.

carlohamalainen committed:
Revert ghc-mod.cabal to current master.

carlohamalainen committed:
Import in alphabetical position.

carlohamalainen committed:
LANGUAGE extensions on one line.

carlohamalainen committed:
Remove unused import.

carlohamalainen committed:
Revert to current master.

carlohamalainen committed:
Remove dependency on System.Process.Streaming

carlohamalainen committed:
Use gmReadProcess. Builds and tests pass with ghc-7.10.3

carlohamalainen committed:
Fix import for Control.Applicative

carlohamalainen committed:
Avoid use of "traverse" which is not in ghc-7.6.3

carlohamalainen committed:
Builds on ghc-7.6.3; added some fallback cases for earlier versions of the GHC API.

carlohamalainen committed:
Move #if things to Language.Haskell.GhcMod.Gap.

carlohamalainen committed:
Add type sig.

carlohamalainen committed:
Remove definition of tdflags (also gets rid of unsafeGlobalDynFlags).
Just use the DynFlags from the current session.

carlohamalainen committed:
Remove constraints on GmOut and GmLog.

carlohamalainen committed:
Fewer imports; move moduleNameToHtmlFile into a where clause.

carlohamalainen committed:
Move filterMatchingQualifiedImport to a where clause; tidy up
indenting.

carlohamalainen committed:
Use parameter of ```gmLog GmDebug``` instead of passing ```""```.

carlohamalainen committed:
Reformat some very long lines.

carlohamalainen committed:
Merge changes from master.

carlohamalainen committed:
Merging changes from master.

carlohamalainen committed:
Tweaks for building on earlier versions of ghc.

carlohamalainen committed:
Adjust ghc version bound on ghcIdeclHiding.

DanielG commented:
So are you done with polishing up the stuff I noted? I.e. should I do another round of review soon or is there still stuff left to do?

carlohamalainen commented:
Got a few more things to fix (e.g. exit code instead of screaming `SUCCESS`). Should be able to get to it over the next day or two.

DanielG commented:
No rush, just give me a shout when you're done :)

carlohamalainen committed:
Tweaks for building on GHC8.

carlohamalainen committed:
Use exit code instead of printing SUCCESS / FAIL.

carlohamalainen committed:
Add ImportedFromSpec to test suite.

carlohamalainen committed:
Use top-level error GMEMissingHaddock when missing haddock html.

carlohamalainen committed:
Use GMError type instead of calling plain "error".

carlohamalainen committed:
Tidying up logging.

carlohamalainen commented:
I'm happy for another review now.

carlohamalainen committed:
Add note about setting documentation:True in .cabal/config.

carlohamalainen committed:
Tweak for resolver on internal symbols (tested with GHC8; added testcase).

carlohamalainen committed:
Forgot "$$".

carlohamalainen committed:
Drop parens in qualified name. New in GHC8?

DanielG commented:
Sorry about the delay, life was getting in the way ;)

This looks very good, I do have a few more points I'd like addressed before this goes into a release but we can merge it anyways.

I'd like to remove the depdency on parsec as it seems parsing those package-ids could trivially done with [splitOn](https://hackage.haskell.org/package/split-0.2.3.1/docs/Data-List-Split.html#v:splitOn) or even [ReadP](https://hackage.haskell.org/package/base-4.9.0.0/docs/Text-ParserCombinators-ReadP.html).

I also think this feature could be more useful if instead of returning a hackage URL straight away it just returned a package-id, module name and binder name. (of course you can still have an option to return a hackage URL)

Lastly I would like the tests to actually check the returned information rather than just checking nothing crashed. If importedFrom were to not return the URL as suggested above I think that would also be a whole lot easier to do.

So I can either merge now and add you as a collaborator so we can keep refining this on master or just keep it here. Your choice.

carlohamalainen committed:
Adding some more test cases.

carlohamalainen committed:
Remove dependency on parsec.

carlohamalainen committed:
Tidyup some warnings.

carlohamalainen committed:
Accidentally removed import of Control.Applicative.

carlohamalainen commented:
I've removed the parsec dependency, now using ReadP.

Can we merge it as it is, add me as a collaborator, and I'll do
1. package-id, module name and binder name;
2. tests check return info

later?

carlohamalainen committed:
Instead of throwing an exception if the haddock file is missing,
show the package and qualified name; module that exports the
symbol, local URL, and Hackage URL.

Example:

    base-4.9.0.0:GHC.Base.Maybe Data.Maybe file:///opt/ghc/8.0.2/share/doc/ghc-8.0.1.20160725/html/libraries/base-4.9.0.0/Data-Maybe.html https://hackage.haskell.org/package/base-4.9.0.0/docs/Data-Maybe.html

carlohamalainen committed:
Improved tests for imported-from command. Actually checks the output
instead of blindly checking if the value was a Right.

carlohamalainen committed:
Merge remote-tracking branch 'remotes/upstream/master' into ghc-imported-from-merge-vs-master

carlohamalainen commented:
I've addressed your comments. Output is now something like

```
    base-4.9.0.0:GHC.Base.Maybe Data.Maybe file:///opt/ghc/8.0.2/share/doc/ghc-8.0.1.20160725/html/libraries/base-4.9.0.0/Data-Maybe.html https://hackage.haskell.org/package/base-4.9.0.0/docs/Data-Maybe.html
```

and the command doesn't throw an exception when the local html file is not present.

Tests also check for actual output as requested.

Good to merge into master now?

DanielG commented:
Actually I have a couple more points:

Separating the fields by spaces isn't going to work since filenames can have those too.

Also outputting two paths seems silly to begin with, how about just returning the "best" one (i.e. local when available and hackage otherwise) and adding two flags, say `--hackage`, `--local` to request a specific one. Then you can split on the first space which is unambiguous. Alternatively you could only return the package/module stuff and let the frontend handle plugging that into the `doc` command (I'd actually prefer that. Just seems more flexible).

Also I think you should use `pkgDoc` (PkgDoc.hs) instead of your `find-module` stuff since that already takes care of abstracting cabal/sandbox/stack away by passing the right package-dbs to ghc-pkg. (Do test that though haven't tried that in a while ;)

Just go ahead and merge it into master (you should be able to do that now), I'll just make an issue so we don't forget to address the remaining stuff before release :)

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+
+    shortcut [ stackGhcPkgFindModule rp mod
+             , hcPkgFindModule       rp mod
+             , _ghcPkgFindModule     rp mod
```
what's up with the underscore?

DanielG replied:
Replace `shortcut` with `runMaybeT . msum . map MaybeT`.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+    -- in @base-4.6.0.1@.
+    -- _ghcPkgFindModule :: String -> IO (Maybe String)
+    _ghcPkgFindModule rp m = do
+        let opts = ["find-module", m, "--simple-output"] ++ ["--global", "--user"] ++ optsForGhcPkg []
```
`optsForGhcPkg [] == []`, also seems to not be called from anywhere else -> dead code.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+    -> m (Maybe String)
+ghcPkgHaddockInterface ghcPkg readProc pkgDbStack p = do
+    hout <- liftIO $ readProc ghcPkg (toHaskellInterfaces p pkgDbStack) ""
+    return $ Safe.lastMay $ words $ reverse . dropWhile (== '\n') . reverse $ hout
```
Either `.` or `$` not both please, also redundant `$` at the end of the line. Maybe you should run `ghc-mod lint` over the source ;)

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+
+    haddockInterfaceFile <- getHaddockInterfaces p'
+
+    case haddockInterfaceFile of
```
`fmap getVisibleExports' <$> getHaddockInterfaces p'` would do just fine. Another three lines saved :)

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+        case iface of
+            Left _          -> throw $ GMEMissingHaddockInterface ifile
+            Right iface'    -> do let m  = map (\ii -> (Haddock.instMod ii, Haddock.instVisibleExports ii)) $ Haddock.ifInstalledIfaces iface' :: [(Module, [Name])]
+                                      m' = map (\(mname, names) -> (showSDoc dflags $ ppr mname, map (showSDoc dflags . ppr) names)) m         :: [(String, [String])]
```
These lines are quite long, >160 chars!

A list comprehension seems more legible:

```
[ (mname,  names)
| ii <- Haddock.ifInstalledIfaces iface'
, let mname = showSDoc dflags $ ppr $ Haddock.instMod ii
      names = map (showSDoc dflags . ppr) $ Haddock.instVisibleExports ii
]
```

Alternatively `do` in the list monad: 

```
do ii <- Haddock.ifInstalledIfaces iface'
   let mname = showSDoc dflags $ ppr $ Haddock.instMod ii
       names = map (showSDoc dflags . ppr) $ Haddock.instVisibleExports ii
   return (mname, names)
```

DanielG replied:
I did point out the 80 char line length guideline before (https://github.com/DanielG/ghc-mod/pull/810#discussion-diff-70889961) that applies everywhere in case that wasn't clear.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+    qualifyName :: [QualifiedName] -> String -> QualifiedName
+    qualifyName qualifiedNames name
+        -- = case filter (postfixMatch name) qualifiedNames of
+        = case nub (filter (name `f`) qualifiedNames) of
```
`nub` is always a red flag. Are you sure a `Set.toList . Set.fromList` wouldn't do? (nub is `O(n^2)`, the set roundtrip `O(n log n)`. I don't know if `n` is likely to be large enough to matter here though.

DanielG replied:
Are you sure you need nub? `Set.toList . Set.fromList` would be `O(n log n)` instead of `O(n^2)`, is `n` likely to be large here?

carlohamalainen replied:
`n` here is the number of names that a module exports; have changed this to use `Set.toList . Set.fromList`.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+refineLeadingDot (MySymbolSysQualified symb)         exports = map (\e -> e { qualifiedExports = f leadingDot e }) exports
+  where
+    leadingDot :: String
+    leadingDot = '.' : last (separateBy '.' symb)
```
Unsafe `last`.

carlohamalainen replied:
Technically this is ok. At some point I want to refactor the types so that these operations aren't so stringy.

``````
commit c995e15277530ad43a5b4aa73e0caab02c89f6f5
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Tue Aug 9 13:05:30 2016 +0800

    Note on use of ```last```.

diff --git a/Language/Haskell/GhcMod/ImportedFrom.hs b/Language/Haskell/GhcMod/ImportedFrom.hs
index 3ee0170..abd230e 100644
--- a/Language/Haskell/GhcMod/ImportedFrom.hs
+++ b/Language/Haskell/GhcMod/ImportedFrom.hs
@@ -535,6 +535,9 @@ refineLeadingDot :: MySymbol -> [ModuleExports] -> [ModuleExports]
 refineLeadingDot (MySymbolUserQualified _)           exports = exports
 refineLeadingDot (MySymbolSysQualified symb)         exports = map (\e -> e { qualifiedExports = f leadingDot e }) exports
   where
+    -- We use leadingDot only when we have an 'MySymbolSysQualified symb' so
+    -- the 'last' will be ok. Sample value of 'symb' in this case is
+    -- "base-4.8.2.0:Data.Foldable.length".
     leadingDot :: String
     leadingDot = '.' : last (splitOn "." symb)
``````

DanielG replied:
I just meant use `lastNote` instead. You would not believe the ammount of time I've spent tracking down `Exception: Prelude.head: empty list` type errors -.-

That invariant might be satisfied now but code changes, better be safe ;)

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+
+-- This definition of separateBy is taken
+-- from: http://stackoverflow.com/a/4978733
+separateBy :: Eq a => a -> [a] -> [[a]]
```
Why not use `splitOn` instead? I'm not sure what license stackoverflow uses for code examples. I think it's CC-SA which is probably going to conflict with the GPL, better not just plain copy stuff from there [until they switch to MIT.](https://meta.stackexchange.com/questions/272956/a-new-code-license-the-mit-this-time-with-attribution-required?cb=1)

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+            qexports       = qualifiedExports mexports -- e.g. ["base-4.8.2.0:GHC.Base.Just", ...]
+        mVisibleExportsMap <- getVisibleExports getHaddockInterfaces pname
+
+        let visibleExportsMap = fromMaybe (throw $ GMEString $ "ImportedFrom: visible exports map is Nothing") mVisibleExportsMap
```
Unless you rely on lazyness here it's probably better to embed this failure into the monadic context since it's there.

DanielG replied:
Actually for this kind of thing I really like just using a pattern on the lhs of `<-` like:

`Just visibleExportsMap <- getVisibleExports getHaddockInterfaces pname`

That will throw a nice error in the monad (using `fail` IIRC) with a sourceref and everything.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+        -- On earlier versions of GHC, our qexports list will not be fully qualified, so it will
+        -- look like ["base:GHC.Base.Just", ...] instead of ["base-4.8.2.0:GHC.Base.Just", ...].
+        -- So if thisModVisibleExports0 is Nothing, fall back to searching on a shorter pname.
+        let thisModVisibleExports = case thisModVisibleExports0 of
```
This doesn't need an explicit `case`, `thisModVisibleExports0 `mplus` Just pname'` would do.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+                                        Just ve -> ve
+                                        Nothing -> let pname' = ((head $ separateBy '-' pname) ++ ":" ++ thisModuleName) in
+                                                        fromMaybe
+                                                            (throw $ GMENoVisibleExports thisModuleName pname')
```
Again, monadic context -> inline that failure.

DanielG replied:
`L.H.GM.Error.liftMaybe` should be applicable here.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+    let symbolToUse :: String
+        symbolToUse = case qnames of
+                        (qq:_) -> qq -- We got a qualified name, with qualified printing. Qualified!
+                        []     -> throw $ GMEString "ImportedFrom: qnames is empty."
```
use `fail` instead of `throw` that embeds it into the monad.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+        extraImportDecls :: [NiceImportDecl]
+        extraImportDecls = case Safe.headMay parsedPackagesAndQualNames of
+            Just (Right (_, x)) -> case moduleOfQualifiedName x of
+                                      Just x' -> [NiceImportDecl
```
This huge whitespace block just looks ugly, can you pull the `NiceImportDecl` thing out into a binder in the enclosing let and use it in here instead please?

DanielG replied:
Also `-XViewPatterns` could save you the second `case`:

```
Just (Right (_, moduleOfQualifiedName -> Just x)) -> ... 
```

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+        successes' :: [(NiceImportDecl, ([String], String))]
+        successes' = mapMaybe toMaybe successes
+
+        stage0 = map (\(m, (e, p)) -> ModuleExports
```
This indent is ugly too, same thing as above. pull it into a binder.

DanielG commented on ghc-mod.cabal:
```diff
@@ -179,6 +181,7 @@ Library
                       , cabal-helper      < 0.8 && >= 0.7.1.0
                       , deepseq           < 1.5
                       , directory         < 1.3
+                      , exceptions        < 0.9
```
Pretty sure you're not using that anymore.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+getVisibleExports getHaddockInterfaces p = do
+    gmLog GmDebug "getVisibleExports" $ strDoc p
+
+    let p' = case splitOn "@" p of
```
Seems like this should be a function, you have an identical block in the function above.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,826 @@
+    optsForGhcPkg ("-no-user-package-conf":rest) = "--no-user-package-conf"        : optsForGhcPkg rest
+    optsForGhcPkg (_:rest) = optsForGhcPkg rest
+
+    runCmd rp cmd opts = liftIO ((Just <$> (rp cmd opts "")) `catch` (\(_::IOError) -> return Nothing))
```
This is no good, you need to get the actual program names/paths from `optPrograms`.

DanielG commented on Language/Haskell/GhcMod/Error.hs:
```diff
@@ -104,6 +103,27 @@ gmeDoc e = case e of
+
+    haddockSuggestion =
+        text "- To generate Haddock docs for dependencies, try:" $$
+            nest 4 (backticks $ text "cabal install --enable-documentation --haddock-hyperlink-source --dependencies-only") $$
```
Isn't it `--only-dependencies`?

DanielG commented:
Phew, long review. We're getting there. It's your bad luck I had too much time on my hands to do a thorough review. Most of the stuff is pretty nitpick though so we ought to be about done. I really have to write some code style guidelines one of these days :)

So never mind the thing about merging early I said before probably best not to loose all those comments.

DanielG commented:
I'm not sure what to do about the `find-module` stuff now, I did see a comment mentioning `pkgDoc` so obviously you were already aware of it. What's wrong with using it over what you're doing there?

carlohamalainen committed:
Don't use Python-style leading underscore.

carlohamalainen committed:
Remove dead code.

carlohamalainen committed:
Use things from Control.Monad.Trans.Maybe.

carlohamalainen committed:
Remove unused dependency.

carlohamalainen committed:
Remove stackoverflow snippet.

Avoid ```fromMaybe (throw ...)``` pattern.

carlohamalainen committed:
Remove another ```fromMaybe (throw ...)``` pattern.

carlohamalainen committed:
fail instead of throw; tidy up some definitions.

carlohamalainen committed:
Use a view pattern.

carlohamalainen committed:
Tidying up.

carlohamalainen committed:
Flag for cabal install is --only-dependencies not --dependencies-only.

carlohamalainen committed:
Avoid use of ```nub```.

carlohamalainen committed:
Note on use of ```last```.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -282,7 +282,7 @@ ghcPkgFindModule mod = do
 
     shortcut [ stackGhcPkgFindModule rp mod
              , hcPkgFindModule       rp mod
-             , _ghcPkgFindModule     rp mod
```
FYI in GHC a leading underscore denotes a typed hole but you can also use it to suppress the "unused symbol" warning when you then go on to define that symbol. I'm not aware of what "python style" you refer to here.

DanielG commented:
@lierdakil tests for 7.10 are failing with

```
0.05s$ stack --no-terminal build --test --no-run-tests
Could not parse './ghc-7.10.3.yaml':
AesonException "Error in $.packages[0]: failed to parse field 'packages': failed to parse field packages: (Invalid package identifier: \".\",\".\")"
```

I have no idea what's going on there.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -552,7 +552,10 @@ refineVisibleExports getHaddockInterfaces exports = mapM f exports
         mVisibleExportsMap <- getVisibleExports getHaddockInterfaces pname
 
-        let visibleExportsMap = fromMaybe (throw $ GMEString $ "ImportedFrom: visible exports map is Nothing") mVisibleExportsMap
+        visibleExportsMap <- case mVisibleExportsMap of
```
I think you misunderstood, there are three kinds of failures:
- A pure exception

```
do let foo = throw "whatever"
   return foo
```
- An excpetion where a monadic value should be

```
do throw "whatever"
```
- An ExceptT/MonadError failure

```
do throwError $ GMEString "whatever"
```

The differences are subtle but in general the first two kinds are annoying since you need to explicitly catch the exceptions whereas the last one just always gets returned as a `Left` by `runGhcModT`.

The difference between 1) and 2) is important though: `throw "whatever" >> return "foo"` will always crash when run, `let foo = throw "whatever" in return (foo, "bar")` will only crash when you touch the fst of that tuple! (lazyness and such)

The GhcModError stuff should always be returned via `throwError` where possible. All you really had to do was to replace `fromMaybe` with `liftMaybe` :)

DanielG replied:
Also `fail` is equivalent to `throwError . GMEString` which is why I suggested you use that before.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -460,7 +460,7 @@ refineAs (MySymbolUserQualified userQualSym) exports = filterM f exports
     f export = do
         -- e.g. "DL"
         case moduleOfQualifiedName userQualSym of
```
I think here a liftMaybe would actually be more concise for example.

DanielG commented on Language/Haskell/GhcMod/ImportedFrom.hs:
```diff
@@ -0,0 +1,825 @@
+
+postfixMatch :: String -> QualifiedName -> Bool
+postfixMatch originalSymbol qName = endTerm `isSuffixOf` qName
+  where endTerm = last $ splitOn "." originalSymbol
```
Another unsafe `last`.

DanielG commented:
Actually let's continue in #823, it's easier if we have a shared branch.
